spec/common/providers/identity.spec.ts

```diff
@@ -85,4 +90,857 @@ describe('identity', () => {
+        'Request has an invalid body.'
+      );
+    });
+
+    it('should error without req body data', () => {
+      const req = ({
+        method: 'POST',
+        header(val: string) {
+          return 'application/json';
+        },
+        body: {},
+      } as unknown) as express.Request;
+
+      expect(() => identity.validRequest(req)).to.throw(
+        'Request has an invalid body.'
+      );
+    });
+
+    it('should error without req body', () => {
+      const req = ({
+        method: 'POST',
+        header(val: string) {
+          return 'application/json';
+        },
+        body: {
+          data: {},
+        },
+      } as unknown) as express.Request;
+
+      expect(() => identity.validRequest(req)).to.throw(
+        'Request has an invalid body.'
+      );
+    });
+
+    it('should not error on valid request', () => {
+      const req = ({
+        method: 'POST',
+        header(val: string) {
+          return 'application/json';
+        },
+        body: {
+          data: {
+            jwt: '1.2.3',
+          },
+        },
+      } as unknown) as express.Request;
+
+      expect(() => identity.validRequest(req)).to.not.throw();
+    });
+  });
+
+  describe('getPublicKey', () => {
+    it('should throw if header.alg is not expected', () => {
+      expect(() => identity.getPublicKey({ alg: 'RS128' }, {})).to.throw(
+        `Provided JWT has incorrect algorithm. Expected ${identity.JWT_ALG} but got RS128.`
+      );
+    });
+
+    it('should throw if header.kid is undefined', () => {
+      expect(() =>
+        identity.getPublicKey({ alg: identity.JWT_ALG }, {})
+      ).to.throw('JWT has no "kid" claim.');
+    });
+
+    it('should throw if the public keys do not have a property that matches header.kid', () => {
+      expect(() =>
+        identity.getPublicKey(
+          {
+            alg: identity.JWT_ALG,
+            kid: '123456',
+          },
+          {}
+        )
+      ).to.throw(
+        'Provided JWT has "kid" claim which does not correspond to a known public key. Most likely the JWT is expired.'
+      );
+    });
+
+    it('should return the correct public key', () => {
+      expect(
+        identity.getPublicKey(
+          {
+            alg: identity.JWT_ALG,
+            kid: '123456',
+          },
+          {
+            '123456': '7890',
+            '2468': '1357',
+          }
+        )
+      ).to.eq('7890');
+    });
+  });
+
+  describe('isAuthorizedCloudFunctionURL', () => {
+    it('should return false on a bad gcf direction', () => {
+      expect(
+        identity.isAuthorizedCloudFunctionURL(
+          `https://us-central1-europe-${PROJECT}.cloudfunctions.net/function-1`,
+          PROJECT
+        )
+      ).to.be.false;
+    });
+
+    it('should return false on a bad project', () => {
+      expect(
+        identity.isAuthorizedCloudFunctionURL(
+          `https://us-central1-${PROJECT}-old.cloudfunctions.net/function-1`,
```

Comment:
What's wrong with this? Because the project and URL don't match?

Reply:
Yes, we expect the url to be in the form `https://${location}-${project}.cloudfunctions.net/...`, but the given url does not match with the given project name, so we should return false.